exploration/semantic-skeletons.md

```diff
@@ -0,0 +1,442 @@
+
+#### DateTime fields
+
+Options:
+
+```
+{$date :datetime dateFields=YMD}  
+{$date :datetime date=YMD}  
+{$date :datetime fields=YMD}
+```
+
+The names `dateFields`, `date`, and `fields` are candidate names for the option that specifies the semantic skeleton string to be used for formatting the date/time value.
+#### TimePrecision
+
+Options:
+```
+{$date :datetime timePrecision=minute}  
+{$date :datetime time=minute}
+```
+(TODO: Add others)
+
+### Design: Use Separate Functions
+
+Some choices:
+
+#### A single `:datetime` function
+
+_Pros_
+- Everything is in one place
+
+_Cons_
+- More combinations of options that can form invalid skeletons
+- Can require verbose placeholders
+
+#### Use separate `:date`, `:time` and `:datetime` functions
+
+_Pros_
+- More tailored to specific user requirements
+- `:date` and `:time` are somewhat self-documenting about the message author's formatting intention
+
+_Cons_
+- Not fully type-safe.
+
+#### Use separate typed functions
+
+`:date`, `:time`, `:datetime`, `:zoneddatetime`, *maybe* `:zoneddate`, `:zonedtime`, `:timezone`
+
+Problem: Most users are likely to prefer date/time/datetime to zoneddate/zonedtime/zoneddatetime
+as the default formatting functions.
+Most "classical" time types are timestamps.
+User intentions might not be met by these names.
+
+_Pros_
+- Helps users get the right results
+- Could be less verbose
+- Better at documenting the message author's intention
+
+_Cons_
+- More functions
+
+
+#### Use separate skeleton functions
+
+Define functions according to the available "field sets" or fields in the semantic skeletons spec.
+Use options to handle field widths.
+Such a list might look like:
+
+- Standalone fields
+  - `:day`, `:weekday`, `:month`, `:year`, `:hour`, `:minute`, `:second`, `:zone`, `:era`
+  - plus the common combo `:time` and `:date`
+- Date Field Sets
+  - `:day-weekday`, `:month-day`, `:month-day-weekday`, `:year-month-day`, `:year-month-day-weekday`
+- Composite field sets
+  - `:date-zone`, `:date-time-zone`, `:time-zone` (note: `time-zone` is two fields, not "timezone")
+
+Examples:
+```
+Your package arrived on {$d :date} at {$d :time}.
+    Your package arrived on Apr 27, 2025 at 10:50 AM.
+Your package arrived on {$d :month-day-weekday} at {$d :time-zone}.
+    Your package arrived on Sunday, April 27th at 10:50 AM PDT.
+Your package arrived on {$d :month-day-weekday month=medium weekday=full} at {$d :time-zone display=full zone=short}.
+    Your package arrive on Sunday, Apr 27th at 10:50 PDT.
+```
+
+
+_Pros_
+- Options focused on field widths
+- Self-documenting: obvious what fields are shown
+- Reserves options for field width
+
+_Cons_
+- _Nineteen_ (or so) functions
+- Names appear generative, but aren't actually
+- Options might be baroque. Some option values might not be compatible with one another.
+  - Expression-wide options (short/medium/long/full) and field-level options might both be needed together
+
```

Comment:
For example:
```js
new Intl.DateTimeFormat('fi', { month: 'long' }).format()
// 'toukokuu'

new Intl.DateTimeFormat('fi', { day: 'numeric', month: 'long' }).format()
// '2. toukokuuta'
```

```suggestion
- Leads to localization failures similar to picture strings,
  as combinations of standalone fields are formatted incorrectly in inflected languages.
```

Reply:
I don't understand this addition. I mean, I understand what you mean, but not why this is specific to this solution. Any skeleton solution should produce the correct output as a result of running the date/time pattern generator on the selected combination of fields and field widths. FWIW, all solutions can produce incorrect results if the information about the desired format is incomplete.

Reply:
My concern is about the affordances we provide. Something like
```
{$d :month}/{$d :day}
```
is much likelier to get actual use than
```
{$d :datetime month=numeric}/{$d :datetime day=numeric}
```
or
```
{$d :date fields=month}/{$d :date fields=day}
```
because it's much shorter, and doesn't immediately communicate to the reader how hacky it is. Meanwhile, the two latter choices hint much more that this isn't good prcatice.

Reply:
Okay, that's a good callout. I think we might make standalone fields less convenient across the board for similar reasons. We enable them, of course, but when you need them they should require more work?

Reply:
I wouldn't say that we "make them less convenient", but more like we don't specifically make them excessively convenient, but this is semantics and I think we agree.

Reply:
```suggestion
- Providing easy-to-use standalone fields may lead to localization failures similar to picture strings,
  as their combinations are formatted incorrectly in inflected languages.
```

Reply:
I'm going to modify this slightly to include the examples for posterity or future use in our own documentation.

```suggestion
- Providing easy-to-use standalone field functions 
  could lead to localization failures similar to picture strings,
  as naive users might use combinations of standalone functions to concatenate date/time values
  that are inappropriate in various other locales and require translator intervention.
> For example, this message hard codes separators, field order, and
> uses standalone representation of the fields:
> ```
> Today is {$d :month} {$d :day}, {$d :year} and it is {$d :hour}:{$d :minute} o'clock.
>```
```

Reply:
Also, note that this is an objection to **any** standalone formats and might be better located in requirements/use cases with a short con for this item (and other designs that bump up against it)

Reply:
Modified suggestion looks good; I edited it to indent the `>` block.